terraform/modules/base/main.tf

```diff
@@ -0,0 +1,115 @@
+	account_name = var.db_account_name
+}
+
+resource "azurerm_service_plan" "main" {
+    name                = "${var.prefix}-cb-todo-app-sp-tf"
+    location            = data.azurerm_resource_group.main.location
+    resource_group_name = data.azurerm_resource_group.main.name
+    os_type             = "Linux"
+    sku_name            = "B1"
+}
+
+resource "azurerm_linux_web_app" "main" {
+    name                = "${var.prefix}-${var.app_service_name}"
+    location            = data.azurerm_resource_group.main.location
+    resource_group_name = data.azurerm_resource_group.main.name
+    service_plan_id     = azurerm_service_plan.main.id
+
+    site_config {
+        application_stack {
+          docker_image_name = "cameronb9/todo-app:prod"
+          docker_registry_url = "https://index.docker.io"
+        }
+    }
+
+    app_settings = {
+		"DOCKER_REGISTRY_SERVER_URL"          = "https://cameronb9"
+		"FLASK_APP"                           = "todo_app/app"
+		"FLASK_ENV"                           = "production"
+		"LOGIN_DISABLED"                      = var.login_disabled
+		"GITHUB_OAUTH_URL"                    = "https://github.com/login/oauth"
+		"HOMEPAGE_URL"                        = "https://${var.prefix}-${var.app_service_name}.azurewebsites.net"
+		"KEY_VAULT_NAME"                      = azurerm_key_vault.main.name
+		"SECRET_KEY"                          = var.secret_key
+		"WEBSITES_ENABLE_APP_SERVICE_STORAGE" = "false"
+		"WEBSITES_PORT"                       = "80"
+    }
+
+	identity {
+		type = "SystemAssigned"
+	}
+}
+
+data azurerm_linux_web_app "main" {
+	name = azurerm_linux_web_app.main.name
+	resource_group_name = data.azurerm_resource_group.main.name
+}
+
+resource "azurerm_key_vault" "main" {
+	name                        = "${var.prefix}-cbt-tf${var.key_vault_rand_str}"
+	location                    = data.azurerm_resource_group.main.location
+	resource_group_name         = data.azurerm_resource_group.main.name
+	tenant_id                   = data.azurerm_client_config.main.tenant_id
+	soft_delete_retention_days  = 7
+	purge_protection_enabled    = false
+	sku_name                    = "standard"
+}
+
+resource "azurerm_key_vault_access_policy" "user" {
+	tenant_id = data.azurerm_client_config.main.tenant_id
+	object_id = data.azurerm_client_config.main.object_id
+	key_vault_id = azurerm_key_vault.main.id
+
+	secret_permissions = ["Get", "List", "Set", "Delete", "Recover", "Purge"]
+}
+
+resource "azurerm_key_vault_access_policy" "main" {
+	tenant_id    = data.azurerm_client_config.main.tenant_id
+	object_id    = data.azurerm_linux_web_app.main.identity[0].principal_id
+	key_vault_id = azurerm_key_vault.main.id
+
+	secret_permissions = ["Get", "List", "Set", "Delete", "Purge"]
+}
+
+
+
+resource "azurerm_key_vault_secret" "github_oauth_client_id" {
+	name = "GITHUB-OAUTH-CLIENT-ID"
+	value = var.github_oauth_client_id
+	key_vault_id = azurerm_key_vault.main.id
+	depends_on = [ azurerm_key_vault_access_policy.user ]
```

Comment:
Is there a particular reason this depends on the access policy? I don't see any harm other than a very small potential delay, but curious!

Reply:
When I was trying to get it to work, it was complaining about the user not having sufficient access to add secrets. I made a lot of changes and moved things around. I'm not entirely sure if this is needed. I'll try remove it and see what happens. 

Currently when I run `terraform apply` locally, it creates the key vault and gives permission to my account to add secrets. When it runs via the pipeline, the service principal gets the permission to add secrets. So if I run it locally, it will then fail when the pipeline runs as it does not get the correct permissions. To get it to work, I had to destroy before running the pipeline. I know we've set this up to run via the pipeline, however, is it worth creating 2 access policies? One for my user and one for the service principal?

Reply:
I removed the `depends_on` and then got this error in the pipeline:

![image](https://github.com/CameronB9/DevOps-Course-Starter/assets/135842865/1e658b4f-10d0-4488-91c8-2cf00fb3d41e)

I assume it's trying to create the secret before the access policy. I added them back in and it works again.

Reply:
Ah of course that makes sense; it always feels a little strange when you're essentially elevating your own permissions, but it can still help us to audit who has/had access; perfectly happy with the conclusion that the `depends_on` makes sense!

On whether to have two access policies, I think that's genuinely a question to think about. In many workplaces, I would consider it very reasonable to give certain individuals direct access so they can at the very least explore issues/debug etc., but if we're tightening our security to its extreme then restricting it solely to the pipeline will enforce everything to happen very publicly and in a process-controlled fashion. If we're going that way, then potentially we'd restrict your user account so you can't run the Terraform locally in the first place!